packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2091,13 +2109,83 @@ export class KeyringController extends BaseController<
       delete state.encryptionSalt;
     });
 
-    this.#password = password;
+    await this.#deriveEncryptionKey(password);
 
     await this.#clearKeyrings();
     await this.#createKeyringWithFirstAccount(keyring.type, keyring.opts);
     this.#setUnlocked();
   }
 
+  /**
+   * Derive the vault encryption key from the provided password, and
+   * assign it to the class variable for later use with cryptographic
+   * functions.
+   *
+   * When the controller has a vault in its state, the key is derived
+   * using the salt from the vault. If the vault is empty, a new salt
+   * is generated and used to derive the key.
+   *
+   * @param password - The password to use for decryption or derivation.
+   * @param options - Options for the key derivation.
+   * @param options.ignoreVaultKeyMetadata - Whether to ignore the vault key metadata
+   */
+  async #deriveEncryptionKey(
+    password: string,
+    options: { ignoreVaultKeyMetadata: boolean } = {
+      ignoreVaultKeyMetadata: false,
+    },
+  ): Promise<void> {
+    this.#assertControllerMutexIsLocked();
+    const { vault } = this.state;
+
+    if (typeof password !== 'string') {
+      throw new TypeError(KeyringControllerError.WrongPasswordType);
+    }
+
+    let salt: string, keyMetadata: unknown;
+    if (vault && !options.ignoreVaultKeyMetadata) {
+      const parsedVault = JSON.parse(vault);
+      salt = parsedVault.salt;
+      keyMetadata = parsedVault.keyMetadata;
+    } else {
+      salt = this.#encryptor.generateSalt();
+    }
+
+    const exportedEncryptionKey = await this.#encryptor.exportKey(
+      await this.#encryptor.keyFromPassword(password, salt, true, keyMetadata),
```

Comment:
Nit: I find it more explicit like this:

```suggestion
      await this.#encryptor.keyFromPassword(password, salt, /* exportable */ true, keyMetadata),
```

Or:

```suggestion
   const exportable = true;
   
   ...
      await this.#encryptor.keyFromPassword(password, salt, exportable, keyMetadata),
```

Anyway, we know this value HAS TO be `true` given that we're exporting it right away, you can omit that comment if you don't like it 😄

Reply:
Yes that makes sense, I also thought about this, but I generally review from GH UI directly, hence I didn't have any type info here 😅 and had to check the method signature.

But that's ok, let's keep it as is! 👍

Reply:
The next time we make a significant breaking change to the browser passworder library / encryptor interface, we should also make it supported an args object / options bag style parameters. Named parameters. Much easier to read that way.